File: app/utils/ui/uigo/horizontaleContainer.lua
local ____lualib = require("lualib_bundle")
local __TS__ArrayIncludes = ____lualib.__TS__ArrayIncludes
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__ArrayForEach = ____lualib.__TS__ArrayForEach
local __TS__ArrayFindIndex = ____lualib.__TS__ArrayFindIndex
local ____exports = {}
local ____observer = require("utils.observer")
local ObservableVal = ____observer.ObservableVal
local camera = require("utils.camera")
local ____utils = require("utils.utils")
local clamp = ____utils.clamp
local in_next_frame = ____utils.in_next_frame
local ____ui_go_utils = require("utils.ui.uigo.ui_go_utils")
local UIGO_ACTION_ID_RELEASED = ____ui_go_utils.UIGO_ACTION_ID_RELEASED
local function newHorizontaleContainer(this_go, size, start_scale)
    if start_scale == nil then
        start_scale = 1
    end
    local check_long_press, recalculate, PLAYBACK, EASING, DURATION_ANIM, array, padding, TIME_LONG_PRESS, selected_obj, _scale, width_child
    function check_long_press(domino)
        timer.delay(
            TIME_LONG_PRESS,
            false,
            function()
                local btn = domino
                if btn == selected_obj.get() and not btn[2].is_long_press then
                    btn[2].is_long_press = true
                    if btn[2].long_press_callback ~= undefined then
                        btn[2].long_press_callback(btn[2].url)
                    end
                end
            end
        )
    end
    function recalculate(new_child)
        local width = width_child + padding
        local area = width * #array
        local half_area = area / 2
        __TS__ArrayForEach(
            array,
            function(____, obj, i)
                local pos = go.get_position(obj[1].url)
                local ____temp_1 = i * width - half_area
                local ____temp_0
                if new_child == undefined then
                    ____temp_0 = width / 2
                else
                    ____temp_0 = 0
                end
                pos.x = ____temp_1 + ____temp_0
                pos.z = 0
                go.animate(
                    obj[1].url,
                    "position",
                    PLAYBACK,
                    pos,
                    EASING,
                    DURATION_ANIM
                )
            end
        )
        if new_child ~= undefined then
            local new_child_pos = go.get_position(new_child[1].url)
            new_child_pos.z = 0
            new_child_pos.x = #array * width - half_area
            go.animate(
                new_child[1].url,
                "position",
                PLAYBACK,
                new_child_pos,
                EASING,
                DURATION_ANIM
            )
        end
        local scale = clamp(size.x / (area + 70), 0, 1.2)
        _scale = scale
        go.animate(
            this_go,
            "scale",
            PLAYBACK,
            vmath.vector3(scale, scale, 1),
            EASING,
            DURATION_ANIM
        )
    end
    PLAYBACK = go.PLAYBACK_ONCE_FORWARD
    EASING = go.EASING_OUTSINE
    DURATION_ANIM = 0.2
    array = {}
    padding = 5
    TIME_LONG_PRESS = 0.3
    selected_obj = ObservableVal(undefined)
    _scale = 1
    go.set_scale(
        vmath.vector3(start_scale, start_scale, 1),
        this_go
    )
    selected_obj.subscribe(function(val)
        if val ~= undefined then
        end
    end)
    local function add_child(child)
        if width_child == undefined then
            width_child = child[1].width()
        end
        go.set_parent(child[1].url, this_go)
        if not __TS__ArrayIncludes(array, child) then
            recalculate(child)
            array[#array + 1] = child
        end
    end
    local function remove(index)
        if array[index + 1] ~= undefined then
            go.delete(array[index + 1][1].url)
            __TS__ArraySplice(array, index, 1)
        end
        recalculate()
    end
    local function clear()
        __TS__ArrayForEach(
            array,
            function(____, v) return go.delete(v[1].url) end
        )
        __TS__ArraySplice(array, 0, #array)
    end
    local function get_index(url)
        return __TS__ArrayFindIndex(
            array,
            function(____, p) return p[1].url == url end
        )
    end
    local delta = vmath.vector3(math.huge, math.huge, 0)
    local function input_update(x, y, action, action_id)
        __TS__ArrayForEach(
            array,
            function(____, obj_)
                if obj_ ~= undefined then
                    local obj = obj_[2]
                    local is_press = obj.is_pressed(x, y, action_id)
                    if is_press then
                        selected_obj.set(undefined)
                        selected_obj.set(obj_)
                    end
                end
            end
        )
        in_next_frame(function()
            if selected_obj.get() ~= undefined then
                local domino = selected_obj.get()
                check_long_press(domino)
                local click_pos = camera.screen_to_world(action.x, action.y)
                local pos = go.get_position(domino[2].url)
                if delta.x == math.huge or delta.y == math.huge then
                    delta.x = click_pos.x - pos.x
                    delta.y = click_pos.y - pos.y
                end
                click_pos.x = click_pos.x - delta.x
                click_pos.y = click_pos.y - delta.y
                click_pos.z = 6
                go.set_position(click_pos, domino[2].url)
                if action.dx ~= 0 or action.dy ~= 0 then
                    if domino[2].on_drag_callback ~= undefined then
                        domino[2].on_drag_callback(domino[2].url, action.dx, action.dy)
                    end
                end
            end
        end)
        if action_id ~= undefined and action_id == UIGO_ACTION_ID_RELEASED then
            if selected_obj.get() ~= undefined then
                selected_obj.get()[2].hovered.set(false)
                selected_obj.get()[2].trigger_on()
                selected_obj.set(undefined)
                delta.x = math.huge
                delta.y = math.huge
            end
        end
    end
    return {
        add_child = add_child,
        remove = remove,
        input_update = input_update,
        get_index = get_index,
        clear = clear,
        recalculate = recalculate,
        array = array,
        size = size
    }
end
return ____exports

File: src/utils/ui/uigo/horizontaleContainer.ts
/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { IObservable, ObservableVal } from "../../../utils/observer";


import * as camera from '../../../utils/camera';


import { v2 } from "./ui_go_util_types";
import { clamp, in_next_frame } from "../../../utils/utils";
import { go_ui_sprite } from "./scalable_container";
import { uigo_button } from "./interfaces/go_ui";
import { UIGO_ACTION_ID_RELEASED } from "./ui_go_utils";



function newHorizontaleContainer<T extends [go_ui_sprite, uigo_button]>(this_go: hash, size: v2, start_scale = 1) {
    const PLAYBACK = go.PLAYBACK_ONCE_FORWARD;
    const EASING = go.EASING_OUTSINE;
    const DURATION_ANIM = 0.2;
    const array: T[] = [];
    const padding = 5;
    const TIME_LONG_PRESS = 0.3;
    const selected_obj = ObservableVal<T | undefined>(undefined);

    let _scale = 1;
    let width_child: number;

    go.set_scale(vmath.vector3(start_scale, start_scale, 1), this_go);

    selected_obj.subscribe((val) => {
        if (val != undefined) {
            //go.set_parent(val[1].id, undefined, true);
        }
    });


    const add_child = (child: T) => {
        if (width_child == undefined)
            width_child = child[0].width();

        go.set_parent(child[0].url, this_go);
        if (!array.includes(child)) {
            recalculate(child);
            array.push(child);
        }
    };

    const remove = (index: number) => {
        if (array[index] != undefined) {

            go.delete(array[index][0].url);
            array.splice(index, 1);
        }
        recalculate();
    };

    const clear = () => {
        array.forEach(v => go.delete(v[0].url));
        array.splice(0, array.length);
    };


    const get_index = (url: hash): number => array.findIndex(p => p[0].url == url);
    const delta = vmath.vector3(math.huge, math.huge, 0);

    const input_update = (x: number, y: number, action: AnyTable, action_id?: string | hash) => {
        array.forEach(obj_ => {
            if (obj_ != undefined) {
                const obj = obj_[1];

                const is_press = obj.is_pressed(x, y, action_id);
                if (is_press) {
                    selected_obj.set(undefined);
                    selected_obj.set(obj_);
                }

            }
        });

        in_next_frame(() => {
            if (selected_obj.get() != undefined) {
                const domino = selected_obj.get() as T;

                check_long_press(domino);

                const click_pos = camera.screen_to_world((action.x), (action.y));
                const pos = go.get_position(domino[1].url);

                if (delta.x == math.huge || delta.y == math.huge) {
                    delta.x = click_pos.x - pos.x;
                    delta.y = click_pos.y - pos.y;
                }

                click_pos.x -= delta.x;
                click_pos.y -= delta.y;

                click_pos.z = 6;
                go.set_position(click_pos, domino[1].url);

                if (action.dx != 0 || action.dy != 0) {
                    if (domino[1].on_drag_callback != undefined)
                        domino[1].on_drag_callback(domino[1].url, action.dx, action.dy);
                }

            }
        });


        if (action_id != undefined && action_id == UIGO_ACTION_ID_RELEASED) {
            if (selected_obj.get() != undefined) {
                (selected_obj.get() as T)[1].hovered.set(false);
                (selected_obj.get() as T)[1].trigger_on();
                selected_obj.set(undefined);
                delta.x = math.huge;
                delta.y = math.huge;
            }
        }
    };

    function check_long_press(domino: T) {
        timer.delay(TIME_LONG_PRESS, false, () => {
            const btn = domino;
            if (btn == selected_obj.get() && !btn[1].is_long_press) {
                btn[1].is_long_press = true;
                if (btn[1].long_press_callback != undefined) {
                    btn[1].long_press_callback(btn[1].url);
                }
            }
        });
    }


    function recalculate(new_child?: T) {
        const width = width_child + padding;
        const area = (width) * array.length;
        const half_area = area / 2;

        array.forEach((obj, i) => {
            const pos = go.get_position(obj[0].url);
            pos.x = i * (width) - half_area + (new_child == undefined ? width / 2 : 0);
            pos.z = 0;

            go.animate(obj[0].url, 'position', PLAYBACK, pos, EASING, DURATION_ANIM);
        });

        if (new_child != undefined) {
            const new_child_pos = go.get_position(new_child[0].url);
            new_child_pos.z = 0;
            new_child_pos.x = array.length * (width) - half_area;
            go.animate(new_child[0].url, 'position', PLAYBACK, new_child_pos, EASING, DURATION_ANIM);
        }
       

        const scale = clamp(size.x / (area + 70), 0, 1.2);
        _scale = scale;
        go.animate(this_go, 'scale', PLAYBACK, vmath.vector3(scale, scale, 1), EASING, DURATION_ANIM);



    }

    return {
        add_child, remove, input_update, get_index, clear, recalculate, array, size
    };
}



